gh-85984: Add POSIX pseudo-terminal functions.

8vasu opened the pull request:
This follows https://github.com/python/cpython/pull/101831. This is one in a series of PRs aimed at cleaning-up, fixing bugs in, introducing new features in, and updating the code in "Lib/pty.py".

This PR answers the following question: `os.forkpty()` and `pty.fork()` return a pair `pid, fd`, where `fd` is a file descriptor of the master end of a pseudo-terminal pair; if at a later stage one needs to make some modifications to the slave end (such as setting termios attributes), then how does one obtain access to it without reimplementing `ptsname()` in Python or loading it from a shared library like someone is doing here: https://stackoverflow.com/questions/52338062/calling-libc-select-from-python-from-pty-master-side?

This is a dependency of https://github.com/python/cpython/pull/101833, which only needs `os.ptsname()`. However, this PR also adds `os.posix_openpt()`, `os.grantpt()`, and `os.unlockpt()` since all of these POSIX functions are "companions" of each other.

Signed-off-by: Soumendra Ganguly [soumendraganguly@gmail.com](mailto:soumendraganguly@gmail.com)

<!-- gh-issue-number: gh-85984 -->
* Issue: gh-85984
<!-- /gh-issue-number -->

8vasu commented on Modules/posixmodule.c:
```diff
@@ -7651,6 +7651,115 @@ os_sched_getaffinity_impl(PyObject *module, pid_t pid)
+[clinic start generated code]*/
+
+static int
+os_posix_openpt_impl(PyObject *module, int oflag)
```
**Test for os.ptsname()**

```
import os
master, slave = os.openpty()
if os.ptsname(master) == os.ttyname(slave):
      print("Success")
else:
      print("Failure")
```
---

**Test for the entire family**

To test these functions, mimic the small example program at the bottom of the POSIX manpage of `posix_openpt()` (https://pubs.opengroup.org/onlinepubs/9699919799/):

```
import os
master_fd = os.posix_openpt(os.O_RDWR|os.O_NOCTTY)
os.grantpt(master_fd)
os.unlockpt(master_fd)
slave_path = os.ptsname(master_fd)
print("slave path:", slave_path)
slave_fd = os.open(slave_path, os.O_RDWR|os.O_NOCTTY)
```
Note that the above is roughly how `openpty()` is implemented on many systems including CPython's fallback `os.openpty()` implementation. This should fail if the `os.unlockpt()` call is commented out. On my Debian 11 system, commenting out the `os.grantpt()` call does **not** result in failure, and this is the expected behavior of the  `grantpt()` C function on my system: please read the next section if you are more interested in this phenomenon.

---

**Note on grantpt() behavior**

From the Linux manpage (https://man7.org/linux/man-pages/man3/grantpt.3.html):

> Many systems implement this function via a set-user-ID helper
       binary called "pt_chown".  On Linux systems with a devpts
       filesystem (present since Linux 2.2), the kernel normally sets
       the correct ownership and permissions for the pseudoterminal
       slave when the master is opened ([posix_openpt(3)](https://man7.org/linux/man-pages/man3/posix_openpt.3.html)), so that
       nothing must be done by grantpt().  Thus, no such helper binary
       is required (and indeed it is configured to be absent during the
       glibc build that is typical on many systems).

To avoid security issues due to the presence of the setuid binary `pt_chown`, my glibc `grantpt()` does not do anything and instead depends on the kernel to set correct ownership and permission attributes of the slave device. Note that `grantpt()` is supposed to set the permissions of the slave to `0620 (crw--w----)`. To demonstrate that the default glibc `grantpt()` on Debian 11 indeed depends on the kernel and does not to anything by itself, try the following.

Let us assume that all steps are being executed by a non-root user. In an interactive Python session, run

```
>>> import os
>>> master_fd = os.posix_openpt(os.O_RDWR|os.O_NOCTTY)
>>> os.ptsname(master_fd)
```

Then go to a different OS shell session and run

```
$ chmod u-w <path-to-slave-from-os-ptsname>
```

Back in Python, run

```
>>> os.grantpt(master_fd)
```

Go back to the OS shell and run

```
$ ls -l <path-to-slave-from-os-ptsname>
```

to see that `os.grantpt()` was unable to change permissions to `0620 (crw--w----)` (but did not fail or raise exceptions).

encukou replied:
Is it possible to automate these?

encukou replied:
(I can help, but I don't know if you've already tried.)

8vasu replied:
Thank you for the review! Apologies for not attending this earlier: I recently moved to a different country.

Correct me if I am wrong, but I do not think there is a file containing tests for `posixmodule.c`. Maybe just put these in a shell script for temporary testing? Do you want me to do this?

8vasu replied:
Edit: Grammar: *... not attending TO this earlier ...

8vasu replied:
I have added a comment about the `O_CLOEXEC` above. I do not remember if I performed these tests after @gpshead made his commits. Therefore, I think I will test again (maybe I will write a shell script to automate these).

encukou replied:
I was thinking of adding these tests to the CI test suite. I'll take a stab at it next week.

8vasu replied:
Thank you so much. Please keep me in the loop so that I can learn how to do this.

encukou replied:
I've added some tests as a commit to this PR. Do they look understandable? Is there anything you'd add?

erlend-aasland commented on configure.ac:
```diff
@@ -4943,6 +4943,21 @@ PY_CHECK_FUNC([setgroups], [
+)
+AC_SEARCH_LIBS([ptsname], [util],
+ [AC_DEFINE([HAVE_PTSNAME], [1], [Define to 1 if you have the `ptsname' function.])]
+)
```
IIRC, AC_SEARCH_LIBS affects the global LIBS. We probably do not want that.

8vasu replied:
Thank you for the review. I had added `os.login_tty()` in https://github.com/python/cpython/pull/29658, where I had originally used `AC_CHECK_FUNCS`. However, @tiran suggested that I use `AC_SEARCH_LIBS` instead, mentioning the following:

> This approach is easier and better. It updates LIBS only when it is necessary and won't add libutils a second time. The login_tty is provided by libutil on Linux, FreeBSD, OpenBSD, and NetBSD.

Should I still go ahead and change them to `AC_CHECK_FUNCS`? Or is there yet another approach that is better?

Edit: fixed grammar

erlend-aasland replied:
You could wrap this whole section with `WITH_SAVE_ENV` just to be sure LIBS (or anything else) is touched. You can grep `configure.ac` for example use.

8vasu replied:
Thank you for your guidance. I will do this now.

8vasu committed:
Add POSIX pseudo-terminal functions.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

8vasu committed:
Merge branch 'main' into posix_pt

gpshead committed:
Merge branch 'main' into posix_pt

gpshead committed:
ReSTify NEWS

gpshead committed:
versionadded:: 3.12

gpshead committed:
Make posix_openpt fd is non-inheritable.

By default all fds returned by Python APIs must be non-inheritable.

encukou committed:
Merge in the main branch

encukou commented on Modules/posixmodule.c:
```diff
@@ -8040,6 +8040,128 @@ os_sched_getaffinity_impl(PyObject *module, pid_t pid)
+
+    sig_saved = PyOS_setsig(SIGCHLD, SIG_DFL);
+    ret = grantpt(fd);
+    PyOS_setsig(SIGCHLD, sig_saved);
```
This relies on the GIL. IMO, that shouldn't block this PR for 3.13, but I'd like to follow up.

@colesbury, I plan to do lots of reviews this year. Is there anything I should do when I spot a case like this?

colesbury replied:
Let's track these in an issue with the "topic-free-threading" label

encukou replied:
Filed here: #114727

encukou commented on Doc/library/os.rst:
```diff
@@ -1386,6 +1399,19 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
    .. versionadded:: 3.3
 
 
+.. function:: posix_openpt(oflag, /)
+
+   Open and return a file descriptor for a master pseudo-terminal device.
+
+   Calls the C standard library function :c:func:`posix_openpt`. The *oflag*
+   argument is used to set file status flags and file access modes as
+   specified in the manual page of :c:func:`posix_openpt` of your system.
```
Should the added `O_CLOEXEC` flag be mentioned here?

8vasu replied:
Looks like @gpshead added that to make the returned master fd non-inheritable. According to his commit:

> By default all fds returned by Python APIs must be non-inheritable.

Given this information, I will let you and @gpshead be the judge and I will make the changes accordingly.

encukou replied:
I'll gladly defer to Greg, but IMO this needs to be the default, and it's not necessary to provide a knob for it (yet?).
But the docs should mention the effect and implementation: the new file descriptor is [non-inheritable]; `O_CLOEXEC` flag is added to *flags*.

encukou committed:
Update whatsnew version

encukou committed:
Add pseudoterminal tests

8vasu committed:
Update os.ptsname() so that it uses the C function ptsname_r() if it is available.
Make os.grantpt() use saved errno upon failure.
Update documentation related to the POSIX pty functions.
Add a test for the POSIX pty functions.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

8vasu committed:
Run /Tools/build/regen-configure.sh

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

encukou commented on Doc/library/os.rst:
```diff
@@ -1429,6 +1443,23 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
+
+   Open and return a file descriptor for a master pseudo-terminal device.
+
+   Calls the C standard library function :c:func:`posix_openpt`. The *oflag*
```
To avoid the docs warnings, C library functions need to go `nitpick_ignore` in `Doc/conf.py`.
Sorry for not catching this earlier. I pushed this directly to save a review round-trip.

encukou committed:
Docs: Add POSIX C functions to nitpick_ignore

encukou reviewed (approved):
+1 from me! 
@gpshead, do you want to take another look?

encukou committed:
Merge branch 'main' into posix_pt
